subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt

```diff
@@ -0,0 +1,528 @@
+typealias ApiTypeSupplier = () -> ApiType
+
+
+internal
+typealias ParameterNamesSupplier = (String) -> List<String>?
+
+
+/**
+ * Provides [ApiType] instances by Kotlin source name from a class path.
+ *
+ * Keeps JAR files open for fast lookup, must be closed.
+ * Once closed, type graph navigation from [ApiType] and [ApiFunction] instances
+ * will throw [IllegalStateException].
+ *
+ * Limitations:
+ * - supports Java byte code only, not Kotlin
+ * - does not support nested Java arrays as method parameters
+ * - does not distinguish co-variance and contra-variance
+ */
+internal
+class ApiTypeProvider(
+    private val repository: ClassBytesRepository,
+    parameterNamesSupplier: ParameterNamesSupplier
+) : Closeable {
+
+    private
+    val context = Context(this, parameterNamesSupplier)
+
+    private
+    val apiTypesBySourceName = mutableMapOf<String, ApiTypeSupplier?>()
+
+    private
+    var closed = false
+
+    fun type(sourceName: String): ApiType? = open {
+        apiTypesBySourceName.computeIfAbsent(sourceName) {
+            repository.classBytesFor(sourceName)?.let { apiTypeFor(sourceName) { it } }
+        }?.invoke()
+    }
+
+    fun allTypes(): Sequence<ApiType> = open {
+        repository.allClassesBytesBySourceName().map { (sourceName, classBytes) ->
+            apiTypesBySourceName.computeIfAbsent(sourceName) {
+                apiTypeFor(sourceName, classBytes)
+            }!!
+        }.map { it() }
+    }
+
+    override fun close() =
+        try {
+            repository.close()
+        } finally {
+            closed = true
+        }
+
+    private
+    fun apiTypeFor(sourceName: String, classBytes: () -> ByteArray) = {
+        ApiType(sourceName, classNodeFor(classBytes), context)
+    }
+
+    private
+    fun classNodeFor(classBytesSupplier: () -> ByteArray) = {
+        ApiTypeClassNode().also {
+            ClassReader(classBytesSupplier()).accept(it, SKIP_DEBUG or SKIP_CODE or SKIP_FRAMES)
+        }
+    }
+
+    private
+    fun <T> open(action: () -> T): T =
+        if (closed) throw IllegalStateException("ApiTypeProvider closed!")
+        else action()
+
+    internal
+    class Context(
+        private val typeProvider: ApiTypeProvider,
+        private val parameterNamesSupplier: ParameterNamesSupplier
+    ) {
+        fun type(sourceName: String): ApiType? =
+            typeProvider.type(sourceName)
+
+        fun parameterNamesFor(typeName: String, functionName: String, parameterTypeNames: List<String>): List<String>? =
+            parameterNamesSupplier("$typeName.$functionName(${parameterTypeNames.joinToString(",")})")
+    }
+}
+
+
+internal
+class ApiType(
+    val sourceName: String,
+    private val delegateSupplier: () -> ClassNode,
+    private val context: ApiTypeProvider.Context
+) {
+
+    val isPublic: Boolean
+        get() = (ACC_PUBLIC and delegate.access) > 0
+
+    val isDeprecated: Boolean
+        get() = delegate.visibleAnnotations.has<java.lang.Deprecated>()
```

Comment:
You're only looking at Java code at this point correct?
I want to make sure you don't also need to look for a groovy `Deprecated` annotation or the Kotlin one.

Reply:
Yes, only Java bytecode